File: 30_days_of_C/C++/Project_19_Bank_Management_System/Bank_Management_System.cpp
#include <iostream>
#include <string>
#include <vector>

class Account {
public:
    Account(int accNumber, std::string accHolder, double balance)
        : accNumber(accNumber), accHolder(accHolder), balance(balance) {}

    int getAccNumber() const { return accNumber; }
    std::string getAccHolder() const { return accHolder; }
    double getBalance() const { return balance; }

    void deposit(double amount) {
        balance += amount;
    }

    bool withdraw(double amount) {
        if (amount > balance) {
            return false;
        }
        balance -= amount;
        return true;
    }

private:
    int accNumber;
    std::string accHolder;
    double balance;
};

class Bank {
public:
    void createAccount(int accNumber, std::string accHolder, double balance) {
        accounts.emplace_back(accNumber, accHolder, balance);
    }

    Account* findAccount(int accNumber) {
        for (auto& account : accounts) {
            if (account.getAccNumber() == accNumber) {
                return &account;
            }
        }
        return nullptr;
    }

    void displayAccountDetails(int accNumber) {
        Account* account = findAccount(accNumber);
        if (account) {
            std::cout << "Account Number: " << account->getAccNumber() << "\n";
            std::cout << "Account Holder: " << account->getAccHolder() << "\n";
            std::cout << "Balance: $" << account->getBalance() << "\n";
        } else {
            std::cout << "Account not found.\n";
        }
    }

private:
    std::vector<Account> accounts;
};

int main() {
    Bank bank;
    int choice;
    do {
        std::cout << "\nBank Management System\n";
        std::cout << "1. Create Account\n";
        std::cout << "2. Deposit Money\n";
        std::cout << "3. Withdraw Money\n";
        std::cout << "4. Display Account Details\n";
        std::cout << "5. Exit\n";
        std::cout << "Enter your choice: ";
        std::cin >> choice;

        int accNumber;
        std::string accHolder;
        double amount;
        switch (choice) {
        case 1:
            std::cout << "Enter Account Number: ";
            std::cin >> accNumber;
            std::cout << "Enter Account Holder Name: ";
            std::cin >> accHolder;
            std::cout << "Enter Initial Balance: ";
            std::cin >> amount;
            bank.createAccount(accNumber, accHolder, amount);
            std::cout << "Account created successfully.\n";
            break;

        case 2:
            std::cout << "Enter Account Number: ";
            std::cin >> accNumber;
            std::cout << "Enter Amount to Deposit: ";
            std::cin >> amount;
            {
                Account* account = bank.findAccount(accNumber);
                if (account) {
                    account->deposit(amount);
                    std::cout << "Amount deposited successfully.\n";
                } else {
                    std::cout << "Account not found.\n";
                }
            }
            break;

        case 3:
            std::cout << "Enter Account Number: ";
            std::cin >> accNumber;
            std::cout << "Enter Amount to Withdraw: ";
            std::cin >> amount;
            {
                Account* account = bank.findAccount(accNumber);
                if (account) {
                    if (account->withdraw(amount)) {
                        std::cout << "Amount withdrawn successfully.\n";
                    } else {
                        std::cout << "Insufficient balance.\n";
                    }
                } else {
                    std::cout << "Account not found.\n";
                }
            }
            break;

        case 4:
            std::cout << "Enter Account Number: ";
            std::cin >> accNumber;
            bank.displayAccountDetails(accNumber);
            break;

        case 5:
            std::cout << "Exiting...\n";
            break;

        default:
            std::cout << "Invalid choice. Please try again.\n";
        }
    } while (choice != 5);

    return 0;
}

File: 30_days_of_python/Project_31_Rock_Paper_Scissor/Project_34_Bank_Management_System/Bank_Management_System.py
# bank_account.py

class BankAccount:
    def __init__(self, account_number, account_name, balance=0):
        self.account_number = account_number
        self.account_name = account_name
        self.balance = balance

    def deposit(self, amount):
        self.balance += amount
        print(f"Deposited ${amount}. New balance: ${self.balance}")

    def withdraw(self, amount):
        if amount > self.balance:
            print("Insufficient balance.")
        else:
            self.balance -= amount
            print(f"Withdrew ${amount}. New balance: ${self.balance}")

    def check_balance(self):
        print(f"Account balance: ${self.balance}")


class BankSystem:
    def __init__(self):
        self.accounts = {}

    def create_account(self, account_number, account_name):
        if account_number in self.accounts:
            print("Account already exists.")
        else:
            self.accounts[account_number] = BankAccount(account_number, account_name)
            print("Account created successfully.")

    def access_account(self, account_number):
        if account_number in self.accounts:
            return self.accounts[account_number]
        else:
            print("Account does not exist.")
            return None


def main():
    bank_system = BankSystem()

    while True:
        print("\n1. Create Account")
        print("2. Access Account")
        print("3. Quit")

        choice = input("Choose an option: ")

        if choice == "1":
            account_number = input("Enter account number: ")
            account_name = input("Enter account name: ")
            bank_system.create_account(account_number, account_name)
        elif choice == "2":
            account_number = input("Enter account number: ")
            account = bank_system.access_account(account_number)
            if account:
                while True:
                    print("\n1. Deposit")
                    print("2. Withdraw")
                    print("3. Check Balance")
                    print("4. Back")

                    choice = input("Choose an option: ")

                    if choice == "1":
                        amount = float(input("Enter amount to deposit: "))
                        account.deposit(amount)
                    elif choice == "2":
                        amount = float(input("Enter amount to withdraw: "))
                        account.withdraw(amount)
                    elif choice == "3":
                        account.check_balance()
                    elif choice == "4":
                        break
                    else:
                        print("Invalid option. Please choose again.")
        elif choice == "3":
            break
        else:
            print("Invalid option. Please choose again.")


if __name__ == "__main__":
    main()
